### PR TITLE
Fix `IS2_INCONSISTENT_SYNC` SpotBugs violations

### PR DESCRIPTION
**Pull request by basil**
Fix [`IS2_INCONSISTENT_SYNC`](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#is-inconsistent-synchronization-is2-inconsistent-sync) SpotBugs violations, mostly by adding `synchronized` keywords to methods that were missing it.

### Proposed changelog entries

N/A

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

File: core/src/main/java/hudson/security/csrf/DefaultCrumbIssuer.java
Comment (basil) [inline core/src/main/java/hudson/security/csrf/DefaultCrumbIssuer.java:60]: Probably only used by XStream, so synchronization is technically unnecessary, but it seems only logical to synchronize given that this method is setting `md` and readers of `md` should not be able to see it in a partially initialized state.

File: core/src/main/java/hudson/slaves/SimpleScheduledRetentionStrategy.java
Comment (basil) [inline core/src/main/java/hudson/slaves/SimpleScheduledRetentionStrategy.java:242]: Everything else that manipulates `lastStart`, `lastStop`, etc. is already synchronized, so this should be synchronized as well for consistency. `updateStartStopWindow` is already synchronized, so it doesn't hurt to synchronize on the whole method.

File: core/src/main/java/hudson/tasks/Fingerprinter.java
Comment (basil) [inline core/src/main/java/hudson/tasks/Fingerprinter.java:395]: `ref` is manipulated elsewhere and is not volatile, so we should synchronize on `this` whenever reading or writing it.

File: core/src/main/java/hudson/util/ProcessTree.java
Comment (basil) [inline core/src/main/java/hudson/util/ProcessTree.java:1730]: `getEnvironmentVariables()` is synchronized, so this should be as well. Not to mention every other platform's `getArguments()` in this class is also synchronized. This seems like an oversight.

File: core/src/main/java/hudson/util/SequentialExecutionQueue.java
Comment (basil) [inline core/src/main/java/hudson/util/SequentialExecutionQueue.java:112]: Yes, the caller must have a lock as the comment states. And callers do have a lock. But since synchronization is re-entrant, it doesn't hurt to explicitly synchronize here, and it also makes SpotBugs happy.

File: core/src/main/java/jenkins/model/RunIdMigrator.java
Comment (basil) [inline core/src/main/java/jenkins/model/RunIdMigrator.java:129]: `idToNumber` is a regular `TreeMap` and is written to by other synchronized methods in this class, so to avoid racing with them we should synchronize on `this` as well.

File: core/src/main/java/jenkins/security/apitoken/ApiTokenStats.java
Comment (basil) [inline core/src/main/java/jenkins/security/apitoken/ApiTokenStats.java:166]: Note the `private synchronized`.

**Review by StefanSpieker — APPROVED** [review]

**Review by res0nance — APPROVED** [review]

**Comment by basil** [comment]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

**Pull request merged**
